### ISSUE TITLE
MicroK8s - Pods with hostNetwork do not have access to cluster DNS

### ISSUE THREAD

**Issue by Thulasiraman1994** [canonical/microk8s#3695]
#### Summary
MicroK8s - Pods with hostNetwork do not have access to cluster DNS

#### What Should Happen Instead?

Pods with hostNetwork should have access to cluster DNS and should get resolved. 
I have even tried with dns policy as - ClusterFirstWithHostNet but still couldn't be able to fix this issue. please find the below details.

#### Reproduction Steps

Create busybox pod with hostnetwork and dns policy as ClusterFirstWithHostNet. 
Create another busy box pod without hostnetwork. 
Create nginx pod and expose it with svc. 

nslookup for my nginx pod fails from hostnetwork pod where it gets resolved as expected from the normal one.

root@master:~# kubectl exec -it busybox -- sh
/ # nslookup nginx
Server: 10.152.183.10
Address 1: 10.152.183.10 kube-dns.kube-system.svc.cluster.local

Name: nginx
Address 1: 10.152.183.139 nginx.default.svc.cluster.local
/ # exit
root@master:~#

root@master:~# kubectl exec -it busybox-hostnet -- sh
/ # nslookup nginx
Server: 10.152.183.10
Address 1: 10.152.183.10

nslookup: can't resolve 'nginx'
/ # exit
command terminated with exit code 1

Pod yamls as below:

apiVersion: v1
kind: Pod
metadata:
name: busybox-hostnet
namespace: default
spec:
containers:

image: busybox:1.28
command:
sleep
"3600"
imagePullPolicy: IfNotPresent
name: busybox
restartPolicy: Always
hostNetwork: true
dnsPolicy: ClusterFirstWithHostNet

coredns yamls:

apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"addonmanager.kubernetes.io/mode":"Reconcile","k8s-app":"kube-dns","kubernetes.io/cluster-service":"true","kubernetes.io/name":"CoreDNS"},"name":"coredns","namespace":"kube-system"},"spec":{"selector":{"matchLabels":{"k8s-app":"kube-dns"}},"strategy":{"rollingUpdate":{"maxSurge":"10%","maxUnavailable":0},"type":"RollingUpdate"},"template":{"metadata":{"annotations":{"priorityClassName":"system-cluster-critical"},"labels":{"k8s-app":"kube-dns"}},"spec":{"containers":[{"args":["-conf","/etc/coredns/Corefile"],"image":"coredns/coredns:1.9.3","imagePullPolicy":"IfNotPresent","livenessProbe":{"failureThreshold":5,"httpGet":{"path":"/health","port":8080,"scheme":"HTTP"},"initialDelaySeconds":60,"successThreshold":1,"timeoutSeconds":5},"name":"coredns","ports":[{"containerPort":53,"name":"dns","protocol":"UDP"},{"containerPort":53,"name":"dns-tcp","protocol":"TCP"},{"containerPort":9153,"name":"metrics","protocol":"TCP"}],"readinessProbe":{"httpGet":{"path":"/ready","port":8181,"scheme":"HTTP"}},"resources":{"limits":{"memory":"170Mi"},"requests":{"cpu":"100m","memory":"70Mi"}},"securityContext":{"allowPrivilegeEscalation":false,"capabilities":{"add":["NET_BIND_SERVICE"],"drop":["all"]},"readOnlyRootFilesystem":true},"volumeMounts":[{"mountPath":"/etc/coredns","name":"config-volume","readOnly":true}]}],"dnsPolicy":"Default","priorityClassName":"system-cluster-critical","serviceAccountName":"coredns","tolerations":[{"key":"CriticalAddonsOnly","operator":"Exists"}],"volumes":[{"configMap":{"items":[{"key":"Corefile","path":"Corefile"}],"name":"coredns"},"name":"config-volume"}]}}}}
  creationTimestamp: "2023-01-25T08:44:54Z"
  generation: 7
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    k8s-app: kube-dns
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: CoreDNS
  name: coredns
  namespace: kube-system
  resourceVersion: "39003"
  uid: c6c9c711-1d6a-44ab-a2f7-607406b9a22e
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: kube-dns
  strategy:
    rollingUpdate:
      maxSurge: 10%
      maxUnavailable: 0
    type: RollingUpdate
  template:
    metadata:
      annotations:
        priorityClassName: system-cluster-critical
      creationTimestamp: null
      labels:
        k8s-app: kube-dns
    spec:
      containers:
      - args:
        - -conf
        - /etc/coredns/Corefile
        image: coredns/coredns:1.9.3
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 5
          httpGet:
            path: /health
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: coredns
        ports:
        - containerPort: 53
          name: dns
          protocol: UDP
        - containerPort: 53
          name: dns-tcp
          protocol: TCP
        - containerPort: 9153
          name: metrics
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /ready
            port: 8181
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            memory: 170Mi
          requests:
            cpu: 100m
            memory: 70Mi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            add:
            - NET_BIND_SERVICE
            drop:
            - all
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/coredns
          name: config-volume
          readOnly: true
      dnsPolicy: Default
      priorityClassName: system-cluster-critical
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: coredns
      serviceAccountName: coredns
      terminationGracePeriodSeconds: 30
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      volumes:
      - configMap:
          defaultMode: 420
          items:
          - key: Corefile
            path: Corefile
          name: coredns
        name: config-volume
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2023-01-25T08:44:54Z"
    lastUpdateTime: "2023-01-25T08:44:57Z"
    message: ReplicaSet "coredns-6f5f9b5d74" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2023-01-25T14:56:40Z"
    lastUpdateTime: "2023-01-25T14:56:40Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 7
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1

#### Introspection Report
Tar ball attached. 

[inspection-report-20230125_153600.tar.gz](https://github.com/canonical/microk8s/files/10500871/inspection-report-20230125_153600.tar.gz)

**Comment by sachinkumarsingh092** [comment]
Hey @Thulasiraman1994, I see this error in the inspection logs:
```
Jan 25 08:45:31 ip-172-31-91-96 microk8s.daemon-kubelite[48068]: }. Err: connection error: desc = "transport: Error while dialing dial unix /var/snap/microk8s/4390/var/kubernetes/backend/kine.sock:12379: connect: connection refused"
.....

Jan 25 09:24:33 master microk8s.daemon-kubelite[729]: W0125 09:24:33.829603     729 endpointslice_controller.go:302] Error syncing endpoint slices for service "kube-system/kube-dns", retrying. Error: EndpointSlice informer cache is out of date
```
Can you share the logs of your coredns and ensure it is working fine?

**Comment by neoaggelos** [comment]
Hi @Thulasiraman1994

See https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy. For pods that use the host network, you also have to explicitly specify the DNS policy as `ClusterFirstWithHostNet` to correctly resolve service and pod names.

**Comment by Thulasiraman1994** [comment]
@neoaggelos I do have the dnspolicy as ClusterFirstWithHostNet, please find the below yaml for your reference. 

root@master:~# kubectl get pods busybox-hostnet -o yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2023-01-25T15:07:25Z"
  name: busybox-hostnet
  namespace: default
  resourceVersion: "259202"
  uid: 037283d7-3c3e-4d71-8240-4cd9410b97c5
spec:
  containers:
  - command:
    - sleep
    - "3600"
    image: busybox:1.28
    imagePullPolicy: IfNotPresent
    name: busybox
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-rgltr
      readOnly: true
  dnsPolicy: ClusterFirstWithHostNet
  enableServiceLinks: true
  hostNetwork: true
  nodeName: worker-1
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kube-api-access-rgltr
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace

**Comment by Thulasiraman1994** [comment]
@sachinkumarsingh092 Coredns is running fine without any errors. Please find the below. 



root@master:~# kubectl logs -f coredns-6f5f9b5d74-6x227 -n kube-system
.:53
[INFO] plugin/reload: Running configuration SHA512 = 591cf328cccc12bc490481273e738df59329c62c0b729d94e8b61db9961c2fa5f046dd37f1cf888b953814040d180f52594972691cd6ff41be96639138a43908
CoreDNS-1.9.3
linux/amd64, go1.18.2, 45b0a11



But I could see a strange behavior here. this is a three-node cluster. If I scale up my coredns across all the worker nodes or when I tried to place the hostnetwork pod on the same node where my coredns is running this is working as expected. this is not the case with my kubeadm cluster. 

In the below example, I have scheduled my hostnet pod on the same node where the coredns is running and now it's getting resolved as expected. Pods with hostnetwork is unable to reach coredns in microk8s even with dnspolicy as ClusterFirstWithHostNet ? 

root@master:~# kubectl get pods -o wide
NAME                     READY   STATUS    RESTARTS       AGE   IP              NODE       NOMINATED NODE   READINESS GATES
nginx-654975c8cd-ml4wx   1/1     Running   0              44h   10.1.226.70     worker-1   <none>           <none>
busybox                  1/1     Running   39 (22m ago)   39h   10.1.133.197    worker-2   <none>           <none>
busybox-hostnet          1/1     Running   0              10s   172.31.92.171   worker-3   <none>           <none>

root@master:~# kubectl get pods -n kube-system -o wide | grep coredns
coredns-6f5f9b5d74-6x227                   1/1     Running   0          39h   10.1.97.200     worker-3   <none>           <none>

root@master:~# kubectl exec -it busybox-hostnet -- sh
/ # nslookup nginx
Server:    10.152.183.10
Address 1: 10.152.183.10 kube-dns.kube-system.svc.cluster.local

Name:      nginx
Address 1: 10.152.183.139 nginx.default.svc.cluster.local
/ # exit

**Comment by neoaggelos** [comment]
Hi @Thulasiraman1994 

In that case, it might be that traffic is dropped by Calico due to a vxlan bug related to packet checksum calculations. Can you test whether the following helps you?

```
microk8s kubectl patch felixconfigurations default --patch '{"spec":{"featureDetectOverride":"ChecksumOffloadBroken=true"}}' --type=merge
```

**Comment by Thulasiraman1994** [comment]
Hi @neoaggelos Thanks a bunch!! It worked as expected!

So, it's an issue from Calico? is this a workaround for all hostnetwork pods ?? any idea when can we expect a fix for this?  Initially we have faced this while installing mayastor through microk8s addon and also while installing mayastor manually in microk8s as we have hostnetwork pods. 
Then I tried it with busybox and nginx to get an understanding about this issue. turns out it is happening from calico itself. is this bug documented somewhere for better understanding?

**Comment by neoaggelos** [comment]
@Thulasiraman1994 this should be happening but it seems like in some situations it doesn't, see #3705 for a fix.

**Comment by datacore-tilangovan** [comment]
@neoaggelos I have tried installing mayastor on a fresh new cluster today with channel=1.26, am still running in to this issue. Running this command manually again fixed the issue. FYI.  

microk8s kubectl patch felixconfigurations default --patch '{"spec":{"featureDetectOverride":"ChecksumOffloadBroken=true"}}' --type=merge

**Comment by neoaggelos** [comment]
Hi @datacore-tilangovan, thanks for reporting.

We are in the process of validating and backporting the fix to stable release channels (1.25 and 1.26).

**Comment by martinblitz** [comment]
FWIW I just ran into this issue in 1.26.1, and this fixed it.  Thank you.
